### DISCUSSION TITLE
Restarting an application within a Windows service - WindowsServiceLifetime

### DISCUSSION THREAD

**Discussion by tjmoore** [dotnet/runtime#71488] [Q&A]
I have an .NET 6 service application that runs can run under a service (Windows or systemd), that has a restore backup function which replaces configs, databases and so on from a backup file and then triggers the application to restart within its process.

This is done using `IHostApplicationLifetime.StopApplication` and using a flag to indicate restart with a loop around RunAsync. If restarting it goes around the loop, preventing the application process from exiting, if not restarting it exits as normal.

For example, the restart loop in Program.cs:

```csharp
do
{
    RestartAftershutdown = false;
    var builder = Host.CreateDefaultBuilder(args);
    ...
    using var host = builder.Build();
    await host.RunAsync();
}
while (RestartAftershutdown);
```

This works fine running in console process, however with `builder.UseWindowsService()` it creates an instance of `WindowsServiceLifetime` and on restart it will throw `InvalidOperationException("Stopped without starting")`. This is in the private Run method in https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs

I think the issue is StopApplication will stop the service whatever even though the process is still running. On start it's telling me the service has stopped obviously.

I don't necessarily want to set the services to restart on error and use a custom implementation to exit with an error code on a non-error restart, especially if this will restart on genuine errors also if they're are not recoverable. Though this might be my only option at the moment. Similar to approaches here https://blog.stephencleary.com/2020/06/servicebase-gotcha-recovery-actions.html. I've tried similar to this to avoid calling `base.OnStop()` in override of OnStop but that doesn't work, the exception still occurs on restart.

Further on using error to restart, I can see this being a bit problem with the recovery options. Set anything other than 0 for days until reset then it will count up errors on each restart. As my installer is WiX, I have three recovery attempt options and not sure what happens after that. If I only set first failure to restart then I would assume I can only restart once until the number of days resets the count? If I set it to 0 and a genuine error occurs, it's going to always restart and risks an infinite restart loop? Just doesn't seem right to use error state to get it to restart.

Not sure what's best here or if there's another approach. I just want to tear down everything and restart from the point of reading configs and starting the application, but not necessarily restarting the host service. Noting that I'm trying to keep an implementation that works on linux also with systemd. Obviously the lifetime host is different there.

**Comment by tjmoore** [comment]
Found possibly similar issue with Linux. UseSystemd() and running under systemd. The restart works fine in the loop and everything restarts, but a minute or two later the service is killed, possibly by systemd and get this in the systemctl log:

```
Jul 01 15:50:04 centos systemd[1]: my-app.service stop-sigterm timed out. Killing.
Jul 01 15:50:04 centos systemd[1]: my-app.service: main process exited, code=killed, status=9/KILL
Jul 01 15:50:04 centos systemd[1]: Unit my-app.service entered failed state.
Jul 01 15:50:04 centos systemd[1]: my-app.service failed.
```

Really in both platforms I want to restart everything internally but not shut down the host, keeping the process. Yet still have the host integration to handle external shutdown & start events.

**Comment by Clockwork-Muse** [comment]
Restarting a service (eg, `sudo systemctl restart <your service>`) I don't believe will trigger any restart counts - those are solely for failure conditions.
Restarting a service _should_, as good practice, assume that the service has been stopped, then started (ie, it's a convenience call).  This is likely why `systemd` kills your service - your service never actually stopped, and the timeout has been hit.

If you have some reason why you don't want a full restart of the service host, don't use the service manager restart - give your individual service a specific message to "reload application" or some such.  But in most cases tying into the service manager's native capabilities is going to be cleaner and better managed.  For instance:

> especially if this will restart on genuine errors also if they're are not recoverable

You have to assume this happens - for instance, if the machine loses power while writing the backup/config files, it's quite possible they're corrupt whenever your service is restarted.  Whether that's manually, or when the machine reboots, you have to assume that anything you read could be in an invalid state.

**Reply by tjmoore** [reply]
I'm not using the service manager to restart, or would prefer to avoid the service manager. Effectively, yes, what I do want is a 'reload application' mechanism, but I can't work out how to do that without using `StopApplication` and reloading the app from the host builder.

The restart process is triggered by the user uploading a backup via a web API and then I call `StopApplication` and intercept the shutdown when restarting by calling the app startup sequence again when it falls out of `RunAsync`, instead of allowing it to exit.

Problem is on Windows `StopApplication` actually stops the host service. Similarly with Linux, systemd seems to kill the process sometime later as it thinks there's a timeout waiting for app to stop. I guess `StopApplication` has signalled it's stopping, but because I don't stop it kills it hard.

Using the hard exit with an error on Windows does work, but requires the service manager set to restart on error, and does trigger a restart count. I don't know with systemd but as said it's killing my process anyway.

There's another way but seems a bit rubbish and have had problems with it also, which is to register a handler on `IApplicationLifetime.ApplicationStopped`, and call the startup methods from there (RunAsync), as that will block and ApplicationStopped itself blocks shutdown. Doing that is stacking up blocked calls each time a restore occurs and not sure of the side-effects of this.

**Reply by tjmoore** [reply]
What would really be great is a new method: `IApplicationLifetime.RestartApplication`

Although depends if people expect that to restart the host, or just the application within the host. Restarting the host programmatically is obviously difficult or impossible if the app doesn't have permissions.

**Reply by davidfowl** [reply]
It sounds like maybe you should build a custom hosting lifetime here. One that has the semantics you want for windows services and for systemd services (if you care about systemd). Are you trying to restart the entire host? If so then maybe it makes sense to create your own derived servicebase class and put the logic in the different events.
